### Logistic Regression Assignment - 91279

In [39]:
import pandas as pd
import numpy as  np
from sklearn import linear_model
from sklearn import preprocessing

In [4]:
# Read titanic train and test data 
train_data = pd.read_csv('train data.csv')
test_data = pd.read_csv('test data.csv')

In [6]:
#Explore the train data
train_data.shape

(891, 12)

In [15]:
train_data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [16]:
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [20]:
# Covert cabin to string
char_cabin = train_data["Cabin"].astype(str)

In [23]:
# Retrieve the first letter of each cabin entry and creating a new numpy array to
# store the value
new_Cabin = np.array([cabin[0] for cabin in char_cabin])

In [24]:
train_data.Cabin = pd.Categorical(new_Cabin)

In [26]:
train_data.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00     n        S  
887  female  19.0      0      0      112053  30.00     B        S  
888  female   NaN      1      2  W./C. 6607  23.45     n        S  
889    male  26.0      0      0      111369  30.00     C        C  
890    male  32.0      0      0      370376   7.75     n        Q

In [36]:
new_age_var = np.where(train_data.Age.isnull(),28,train_data.Age)

In [37]:
train_data.Age = new_age_var

In [38]:
#train_data.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00     n        S  
887  female  19.0      0      0      112053  30.00     B        S  
888  female  28.0      1      2  W./C. 6607  23.45     n        S  
889    male  26.0      0      0      111369  30.00     C        C  
890    male  32.0      0      0      370376   7.75     n        Q

In [40]:
#Initialize label encoder
label_encoder = preprocessing.LabelEncoder()

In [47]:
#Convert sex feature to numeric, logistic reression only understandds numeric values
encoded_sex = label_encoder.fit_transform(train_data.Sex)

In [48]:
encoded_sex

array([1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0,
       0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1,
       0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0,
       1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1,
       0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0,
       1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1,
       0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0,

In [43]:
log_model = linear_model.LogisticRegression()

In [44]:
log_model.fit(X=pd.DataFrame(encoded_sex),
             y= train_data.Survived)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [45]:
log_model.coef_  

array([[-2.43010712]])

In [46]:
log_model.intercept_

array([1.00027876])

In [57]:
predictions = log_model.predict_proba(X=pd.DataFrame(encoded_sex))

In [63]:
predictions = pd.DataFrame(data=predictions,columns=["Death_Probability","Survival_Probability"])

In [64]:
predictions.head()

Death_Probability  Survival_Probability
0           0.806875              0.193125
1           0.268887              0.731113
2           0.268887              0.731113
3           0.268887              0.731113
4           0.806875              0.193125

In [69]:
#Generate survival predictions vs Sex
pd.crosstab(train_data.Sex, predictions.loc[:,"Death_Probability"])

Death_Probability  0.268887  0.806875
Sex                                  
female                  314         0
male                      0       577

In [70]:
pred = log_model.predict(pd.DataFrame(encoded_sex))

In [73]:
#Add more features to the model
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500     n        S  
1      0          PC 17599  71.2833     C        C  
2      0  STON/O2. 3101282   7.9250     n        S  
3      0            113803  53.1000     C        S  
4      0            373450   8.0500     n        S

In [74]:
encoded_class = label_encoder.fit_transform(train_data.Pclass)

In [75]:
encoded_cabin = label_encoder.fit_transform(train_data.Cabin)

In [86]:
train_features = pd.DataFrame([encoded_class,encoded_cabin,encoded_sex,train_data.Age]).T

In [87]:
train_features.head()

0    1    2     3
0  2.0  8.0  1.0  22.0
1  0.0  2.0  0.0  38.0
2  2.0  8.0  0.0  26.0
3  0.0  2.0  0.0  35.0
4  2.0  8.0  1.0  35.0

In [88]:
log_model.fit(X=train_features,y=train_data.Survived)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [90]:
print(log_model.intercept_)
print(log_model.coef_)

[3.32716302]
[[-0.90790164 -0.06426483 -2.43179802 -0.0265924 ]]


In [91]:
pred = log_model.predict(train_features)
pd.crosstab(pred,train_data.Survived)

Survived    0    1
row_0             
0         463   98
1          86  244

In [94]:
#The confusion matrix
#TP = 244, TN = 463, FP=86, FN=98

In [97]:
#Accuracy of the model
accuracy = (463+244)/len(train_data) *100

In [102]:
print("Accuracy:",accuracy,"%")

Accuracy: 79.34904601571269 %


In [103]:
log_model.score(X=train_features,y=train_data.Survived)

0.7934904601571269

In [104]:
print(244/(244+98)*100)

71.34502923976608


In [105]:
from sklearn import metrics

In [107]:
#View Confusion Matrix
metrics.confusion_matrix(y_true=train_data.Survived, y_pred=pred)

array([[463,  86],
       [ 98, 244]], dtype=int64)

In [108]:
print(metrics.classification_report(y_true=train_data.Survived, y_pred=pred))

             precision    recall  f1-score   support

          0       0.83      0.84      0.83       549
          1       0.74      0.71      0.73       342

avg / total       0.79      0.79      0.79       891



In [109]:
test_data = pd.read_csv('test data.csv')

In [110]:
test_data.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [112]:
char_cabin = test_data["Cabin"].astype(str)     # Convert cabin to str

new_Cabin = np.array([cabin[0] for cabin in char_cabin]) # Take first letter

test_data["Cabin"] = pd.Categorical(new_Cabin)  # Save the new cabin var

# Impute median Age for NA Age values
new_age_var = np.where(test_data["Age"].isnull(), # Logical check
                       28,                       # Value if check is true
                       test_data["Age"])      # Value if check is false

test_data["Age"] = new_age_var 

In [113]:
# Convert test variables to match model features
encoded_sex = label_encoder.fit_transform(test_data["Sex"])
encoded_class = label_encoder.fit_transform(test_data["Pclass"])
encoded_cabin = label_encoder.fit_transform(test_data["Cabin"])

test_features = pd.DataFrame([encoded_class,
                              encoded_cabin,
                              encoded_sex,
                              test_data["Age"]]).T

In [114]:
test_preds = log_model.predict(X=test_features)

In [115]:
test_preds

array([0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1,
       1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0,
       1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1,
       0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,

## Questions

### 1. Role of the sigmoid function defined in Cell number 3 in the tutorial

The sigmoid function which is also referred to as the logistic function is used to map a predicted value to a probability between 0 and 1. The function takes a value (t) in this case a numpy array as the input and maps the values of the array to a probability by using the function 1/(1+e^(-t) where t is the output of a linear regression function.

In [118]:
def sigmoid(t):
    return (1/(1 + np.e**(-t)))    

In [133]:
#The sigmoid function converts the input value to a probability value between 0 and 1
print(sigmoid(1.111))

0.24768450465760494


In [132]:
print(sigmoid(np.arange(-6,6,0.1)))

[0.00247262 0.00273196 0.00301842 0.00333481 0.00368424 0.00407014
 0.00449627 0.0049668  0.0054863  0.0060598  0.00669285 0.00739154
 0.00816257 0.0090133  0.0099518  0.01098694 0.01212843 0.01338692
 0.01477403 0.0163025  0.01798621 0.01984031 0.02188127 0.02412702
 0.02659699 0.02931223 0.03229546 0.03557119 0.03916572 0.04310725
 0.04742587 0.05215356 0.05732418 0.06297336 0.06913842 0.07585818
 0.0831727  0.09112296 0.09975049 0.10909682 0.11920292 0.13010847
 0.14185106 0.15446527 0.16798161 0.18242552 0.19781611 0.21416502
 0.23147522 0.24973989 0.26894142 0.2890505  0.31002552 0.33181223
 0.35434369 0.37754067 0.40131234 0.42555748 0.450166   0.47502081
 0.5        0.52497919 0.549834   0.57444252 0.59868766 0.62245933
 0.64565631 0.66818777 0.68997448 0.7109495  0.73105858 0.75026011
 0.76852478 0.78583498 0.80218389 0.81757448 0.83201839 0.84553473
 0.85814894 0.86989153 0.88079708 0.89090318 0.90024951 0.90887704
 0.9168273  0.92414182 0.93086158 0.93702664 0.94267582 0.9478

In [128]:
#The np.arange is used to generate an array containing values between the start value and stop value exluding the stop value The step refers to the interval.
np.arange(1,5,2)

array([1, 3])

### 2. When does a sigmoid function output a probability greater than 0.5? 

The sigmoid function outputs a value greater than 0.5 when the output of the linear regression function is a positive value;i.e. the linear combination of variables produces positive numbers

In [144]:
#Example; say we have a linear regression function : y = 2 + 0.5x
#Output is positive
t = 2+0.5*4
print("Positive Value:",t)
print("Probability:", sigmoid(t))

Positive Value: 4.0
Probability: 0.9820137900379085


In [145]:
#Output is negative
t = 2+0.5*-5
print("Negative Value:",t)
print("Probability:", sigmoid(t))

Negative Value: -0.5
Probability: 0.3775406687981454


### 3. Age is a feature in the titanic data. It has some missing values:
#### a.) Explain how the missing values are handled in the tutorial 

The missing values are replaced with the median age of the data set which is 28.

#### b.) Use a panda function to imputing the missing values in age, to obtain the same results as in the tutorial 

In [170]:
data = pd.read_csv('train data.csv')

In [171]:
#Computing the total number of null values
data.Age.isnull().sum()

177

In [172]:
data.Age.median()

28.0

In [173]:
#Replacing the null values with the median of age
new_age = data.Age.fillna(data.Age.median())

In [174]:
#Replace the old age values with the new values
data.Age = new_age

In [178]:
#No null values
data.Age.isnull().sum()

0

### 4. Do sklearn’s machine learning functions allow for the features to be categorical?

SK Learn machine learning functions only deal with real numbers

### 5. In the tutorial, Label Encoding is used 
#### a. Explain the importance of label encoding 

Label Encoding is used to convert categorical values to real numbers so that the different algorithms can understand the feature.

#### b.Describe how it has been implemented in the tutorial

Label encoding has been used to convert __Sex__ which contains categorical values to real numbers, initally the values were M and F but after applying encoding the values were converted to 0 and 1 respectivley. 0 represents male and 1 represents female.

In [182]:
data.Sex.head()

0      male
1    female
2    female
3    female
4      male
Name: Sex, dtype: object

In [183]:
encoded_sex = label_encoder.fit_transform(data.Sex)

In [185]:
#After encoding, replace the original values
data.Sex = encoded_sex
data.Sex.head()

0    1
1    0
2    0
3    0
4    1
Name: Sex, dtype: int64

#### c. Describe other ways in which encoding could have been done

A problem with label encoding is that since there is no relationship between the rows, the algorithm may misunderstand and conclude that 1>0; therefore female > male. *One Hot Encoder* solves this problem. It takes a column which has categorical data, which has been label encoded, and then splits the column into multiple columns. The numbers are replaced by 1s and 0s, depending on which column has what value.

### 6. What is the logistic regression model learned?

In [190]:
print("Intercept :",log_model.intercept_,"\n" "Coefficents:",log_model.coef_)

Intercept : [3.32716302] 
Coefficents: [[-0.90790164 -0.06426483 -2.43179802 -0.0265924 ]]


z = 3.32716302 +  -0.90790164**x1** + -0.06426483**x2** + -2.43179802**x3** + -0.0265924**x4**
* where :
    * x1 = Class
    * x2 = Cabin
    * x3 = Sex
    * x4 = Age

### 7. Discuss the difference between the methods predict and predict_proba of a logistic regression model 

Predict returns the class the predicted value belongs to; whether a 0 and 1, whereas predict_proba returns the probabilities of predicted value to belong to either classes. 
* **Predict** - get the predicted classes
* **predict_proba()** - get the predicted class probabilities

### 8. In regards to metrics: 
#### a. What is a confusion matrix?

A confusion matrix is a table that is used to describe the performance of a classification model.

#### b. Discuss the confusion matrix of the model learned 

In [193]:
pd.crosstab(pred,train_data.Survived)

Survived    0    1
row_0             
0         463   98
1          86  244

* **True Positive** (244) -Refers to the number of people the model predicted to survive and they actually survived.
* **True Negative** (463) -Refers to the number of people the model predicted to not survve and they did not survive.
* **False Positive** (86) -Refers to the number of people the model predicted to survive and they did not survive.
* **False Negative** (98) -Refers to the number of people the model predicted to not survive and they actually survived.

In [ ]:
titani